File: Language/C#/Book/tmp/LINQ实战/Chapter13.CSharp - LinqBooks/LinqBooks.Web/AppCommon.cs
#region Imports

using System;
using System.Data;
using System.Configuration;
using System.Linq;
using System.Web;
using System.Web.Security;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;

using LinqBooks.Entities;

#endregion Imports

namespace LinqBooks.Web
{
  static class AppCommon
  {
    public static User User
    {
      get
      {
        MembershipUser membershipUser = Membership.GetUser();
        if (membershipUser == null)
          throw new Exception("No current user.");

        Guid userId = (Guid)membershipUser.ProviderUserKey;
        if (userId.Equals(HttpContext.Current.Application["UserId"]))
          return (User)HttpContext.Current.Application["User"];

        LinqBooksDataContext dataContext = new LinqBooksDataContext();
        User user = dataContext.Users.SingleOrDefault(u => u.ID == userId);
        if (user == null)
          throw new Exception("Current user not recognized by the application.");

        HttpContext.Current.Application["UserId"] = userId;
        HttpContext.Current.Application["User"] = user;
        return user;
      }
    }
  }
}

File: Language/C#/Book/tmp/LINQ实战/Chapter13.VB - LinqBooks/LinqBooks.Web/AppCommon.vb
Imports LinqBooks.Entities

Module AppCommon

  Public ReadOnly Property User() As User
    Get
      Dim membershipUser = Membership.GetUser()
      If membershipUser Is Nothing Then
        Throw New Exception("No current user.")
      End If

      Dim userId = CType(membershipUser.ProviderUserKey, Guid)
      If userId.Equals(HttpContext.Current.Application("UserId")) Then
        Return CType(HttpContext.Current.Application("User"), User)
      End If

      Dim dataContext = New LinqBooksDataContext()
      Dim result As User = dataContext.Users.SingleOrDefault(Function(u) u.ID = userId)
      If result Is Nothing Then
        Throw New Exception("Current user not recognized by the application.")
      End If

      HttpContext.Current.Application("UserId") = userId
      HttpContext.Current.Application("User") = result
      Return result
    End Get
  End Property

End Module